spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java

```diff
@@ -350,8 +355,24 @@ public void stop(Runnable callback) {
 				public void handle(Exception thrownException, final ConsumerRecord message) {
 					final byte[] key = message.key() != null ? Utils.toArray(ByteBuffer.wrap((byte[]) message.key()))
 							: null;
-					final byte[] payload = message.value() != null
+					byte[] payload = message.value() != null
 							? Utils.toArray(ByteBuffer.wrap((byte[]) message.value())) : null;
+					if (properties.getHeaderMode().equals(HeaderMode.embeddedHeaders)) {
+						EmbeddedHeadersMessageConverter embeddedHeadersMessageConverter = new EmbeddedHeadersMessageConverter();
+						try {
+							MessageValues messageValues = embeddedHeadersMessageConverter.extractHeaders(MessageBuilder.withPayload(payload).build(), true);
```

Comment:
Wouldn't it be better if we overload the method instead of changing the entire signature, which could create compatibility problems?